File: back/src/test/java/com/back/domain/book/service/BookViewsServiceTest.java
package com.back.domain.book.service;

import com.back.domain.book.dto.BookDto;
import com.back.domain.book.entity.Book;
import com.back.domain.review.service.ReviewService;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.webmvc.test.autoconfigure.AutoConfigureMockMvc;
import org.springframework.test.context.ActiveProfiles;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

import static org.assertj.core.api.AssertionsForClassTypes.assertThat;

@ActiveProfiles("test")
@SpringBootTest
@Transactional
public class BookViewsServiceTest {

    @Autowired
    private BookService bookService;

    @Test
    @DisplayName("조회수 기반 도서 순위 조회")
    void t1() {

        List<Integer> viewsCount = List.of(3, 5, 1);

        for (int i = 0; i < viewsCount.size(); i++) {
            for (int j = 0; j < viewsCount.get(i); j++) {
                Book book = bookService.getBookById(i + 1);
                bookService.incrementViewCount(book);
            }
        }

        List<Book> bookRank = bookService.getBooksOrderByRank("views", 0, 10);

        int upperCnt = Integer.MAX_VALUE;

        for (Book book : bookRank) {
            int nowCnt = bookService.getBookViewCount(book);

            System.out.println("view: " + nowCnt);

            assertThat(nowCnt).isEqualTo(viewsCount.get((int) (book.getId() - 1)));
            assertThat(upperCnt).isGreaterThanOrEqualTo(nowCnt);

            upperCnt = nowCnt;

        }

    }
}

File: back/src/test/java/com/back/domain/book/service/BookViewsServiceTest.kt
package com.back.domain.book.service

import com.back.domain.book.entity.Book
import org.assertj.core.api.AssertionsForClassTypes
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.test.context.ActiveProfiles
import org.springframework.transaction.annotation.Transactional

@ActiveProfiles("test")
@SpringBootTest
@Transactional
class BookViewsServiceTest {
    @Autowired
    private lateinit var bookService: BookService

    @Test
    @DisplayName("도서 인기순(조회수) 다건 조회")
    fun t1() {
        val viewsCount = listOf(3, 5, 1)

        for (i in viewsCount.indices)
            (0..<viewsCount[i]).forEach { _ ->
                val book = bookService.getBookById((i + 1).toLong())
                bookService.incrementViewCount(book)
            }

        val bookRank: List<Book> = bookService.getBooksOrderByRank("views", 0, 10)

        var upperCnt = Int.MAX_VALUE

        for (book in bookRank) {
            val nowCnt = bookService.getBookViewCount(book)

            AssertionsForClassTypes.assertThat(nowCnt).isEqualTo(viewsCount[(book.id - 1).toInt()])
            AssertionsForClassTypes.assertThat(upperCnt).isGreaterThanOrEqualTo(nowCnt)

            upperCnt = nowCnt
        }
    }
}
